book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,342 @@
+fine (and encouraged) for caller code to unwrap the expected variants with
+`unreachable!` on the unexpected variants. This is slightly inconvenient but
+it means that we have a unified state interface with unified backpressure.
+
+This API includes both write and read calls. Spotting `Commit` requests in
+code review should not be a problem, but in the future, if we need to
+restrict access to write calls, we could implement a wrapper service that
+rejects these, and export "read" and "write" frontends to the same inner service.
+
+### `Request::CommitBlock(Arc<Block>)`
+[request-commit-block]: #request-commit-block
+
+Performs contextual validation of the given block, committing it to the state
+if successful. Returns `Response::Added(BlockHeaderHash)` with the hash of
+the newly committed block or an error.
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.
+
+Otherwise, attempt to perform contextual validation checks and the commit the
+given block to the state. The exact list of contextual validation checks will
+be specified in a later RFC. If contextual validation checks succeed, the new
+block is added to one of the in-memory chains. If the resulting chain is
+longer than 100 blocks, the oldest block is now past the reorg limit, so we
+remove it from all in-memory chains and commit it to sled as described below
+in `CommitFinalizedBlock`.
+
+Finally, process any queued children of the newly committed block the same way.
+
+### `Request::CommitFinalizedBlock(Arc<Block>)`
+[request-commit-finalized-block]: #request-finalized-block
+
+Commits a finalized block to the sled state, skipping contextual validation.
+This is exposed for use in checkpointing, which produces in-order finalized
+blocks. Returns `Response::Added(BlockHeaderHash)` with the hash of the
+committed block if successful.
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.  Otherwise, call the wrapper function described below, then
+process any queued children.  (Although the checkpointer generates verified
+blocks in order when it completes a checkpoint, the blocks are committed in the
+response futures, so they may arrive out of order).
+
+Committing a block to the sled state should be implemented as a wrapper around
+a function also called by [`Request::CommitBlock`](#request-commit-block),
+which should:
+
+1. Obtain the highest entry of `hash_by_height` as `(old_height, old_tip)`.
+Check that `block`'s parent hash is `old_tip` and its height is
+`old_height+1`, or panic.  This check is performed as defense-in-depth
+to prevent database corruption, but it is the caller's responsibility to
+commit finalized blocks in order.
+
+2. Insert:
+    - `(hash, height)` into `height_by_hash`;
+    - `(height, hash)` into `hash_by_height`;
+    - `(height, block)` into `block_by_height`.
+
+3.  Update the `sprout_anchors` and `sapling_anchors` trees with the Sprout
```

Comment:
If (Sapling|Blossom) pre-Heartwood, we can parse the `root_bytes` field in the `block::Header` as a `sapling::tree::Root` (`hashFinalSaplingRoot`) and insert it into the `sapling::anchors` tree. 

But for everything else, including all the Sprout anchors, we have to compute the whole note commitment tree after applying the transactions in-order. We have to do this already for Sprout in-between `JoinSplit`s, but for Sapling, we only have to do it at the end, all the transactions without `JoinSplit`s can be validated without computing the tree. 

For Sapling beyond Blossom, we could compute the final merkle tree and its root in here, but it's completely non-contextual (completely contained within the block) so I think it should be done in the `BlockVerifier`. This also dovetails with the Sprout work which _has_ to be done in the `TransactionVerifier` anyway.

So! We have a bunch of block metadata that we have  pre-computed and want to save down along with the block we are finalizing that it is derived from. Maybe with an extra argument in the request?

`Request::CommitFinalizedBlock(Arc<Block>, TreeState(Vec<Nullifier>, tree::Root))`

Then for (Sapling|Blossom) we'd also check that the `tree::Root` matches the `root_bytes` defined in the `block::Header` to make sure we computed it correctly.

Reply:
Adding extra data in the request seems good.  I don't think we need to pass the nullifiers in, because they're already in the block data and there's no extra computation required to pull them out.  It sounds like, regardless of whether or not the Sapling root hash is present in the block, fully verifying a block requires computing both the Sprout and Sapling root hashes.  So we could pass them in as

```rust
enum Request {
    CommitBlock {
        block: Arc<Block>,
        sprout_anchor: sprout::tree::Root,
        sapling_anchor: sapling::tree::Root,
    },
    CommitFinalizedBlock {
        block: Arc<Block>,
        sprout_anchor: sprout::tree::Root,
        sapling_anchor: sapling::tree::Root,
    }
}
```

Then the block verifier can pass the values it computed already (saving a recomputation) and the checkpoint verifier can compute the values and pass them in. Either way we don't have to compute anything more than once.

Reply:
Sounds good! 🕺

Reply:
OK, this sounds like a plan.  There's a slight obstacle which is that currently I don't think that we have code on hand to compute these things.  But after making this change, we'll need the checkpoint verifier to fill in some values.  It seems like there are two options:

1. Implement computation of sprout and sapling roots before making the sled changes, then compute the values and fill them into the request;

2. Just fill in zeros until later, and watch out for surprise breakage of other checks that happens as a consequence of filling in garbage values into the database.  Then later, implement the correct logic and wipe the database.

Reply:
I'd prefer it if we didn't substitute invalid data - it makes it much harder to diagnose verification errors in unrelated changes. (Or in newer code that manages to successfully re-use older versions of the database - until it fails later on.)

Instead, let's either:
1) Create the final state service interface, but ignore any values passed in, and don't commit them to sled
2) Create an interim state service interface without the missing values

Reply:
I'd also prefer not to substitute invalid data.  What if we implement computation of sprout and sapling roots now, so they will be ready by the time we finish this design and start trying to implement it?

Reply:
I've started working on this 🎄

Reply:
This plan won't work because recent changes made the checkpointer feed out of order blocks without updating the state code, so things are now broken:
```
Message:  assertion failed: `(left == right)`
  left: `block::Hash("376d45184781a016adf8be512fe60f500d4b687aa751be84a9e594c8c9840100")`,
 right: `block::Hash("08ce3d9731b000c08338455c8a4a6bd05da16e26b11daa1b917184ece80f0400")`: out of order block: the next block must be a child of the current tip
```
I will implement an interim fix.

Reply:
@hdevalence is this comment good to resolve?